traits/ctraits.c

```diff
@@ -2970,22 +2970,71 @@ trait_traverse(trait_object *trait, visitproc visit, void *arg)
     return 0;
 }
 
+/*-----------------------------------------------------------------------------
+|  Identify double underscore names like "__qualname__" and "__package__".
+|
+|  Returns 1 if name starts and ends with "__", else 0. Returns -1 with
+|  an exception set on failure.
++----------------------------------------------------------------------------*/
+
+static int
+is_dunder_name(PyObject *name)
+{
+    Py_ssize_t name_length;
+    int kind;
+    void *data;
+
+    if (PyUnicode_READY(name) < 0) {
+        return -1;
+    }
+
+    name_length = PyUnicode_GET_LENGTH(name);
+    kind = PyUnicode_KIND(name);
+    data = PyUnicode_DATA(name);
+    return (
+        (name_length >= 2)
+        && (PyUnicode_READ(kind, data, 0) == '_')
+        && (PyUnicode_READ(kind, data, 1) == '_')
+        && (PyUnicode_READ(kind, data, name_length - 2) == '_')
+        && (PyUnicode_READ(kind, data, name_length - 1) == '_')
+    );
+}
+
 /*-----------------------------------------------------------------------------
 |  Handles the 'getattr' operation on a 'CTrait' instance:
 +----------------------------------------------------------------------------*/
 
 static PyObject *
 trait_getattro(trait_object *obj, PyObject *name)
 {
-    PyObject *value = PyObject_GenericGetAttr((PyObject *)obj, name);
+    PyObject *value;
+    int is_dunder;
+
+    value = PyObject_GenericGetAttr((PyObject *)obj, name);
     if (value != NULL || !PyErr_ExceptionMatches(PyExc_AttributeError)) {
         return value;
     }
-
     PyErr_Clear();
 
-    Py_INCREF(Py_None);
-    return Py_None;
+    /* The attribute lookup failed. For backwards compatibility, we return
+       `None` in this case, *except* for attribute names that start and end
+       with a double underscore. Those attribute names may have special meaning
+       to the interpreter, and `None` may not be a valid value, so we allow the
+       exception to propagate. */
+
+    is_dunder = is_dunder_name(name);
+    if (is_dunder < 0) {
+        return NULL;
```

Comment:
Are we being defensive here when we handle the case where `name` isn't in the "canonical" representation? I'm not even sure what "canonical" representation means. Does it mean "invalid" unicode?

Reply:
Yes, we're being defensive, possibly unnecessarily so. 

For background on what canonical means, see: [PEP 393](https://www.python.org/dev/peps/pep-0393). Essentially, some deprecated, legacy APIs can produce `Py_Unicode` instances that don't have all of the expected fields initialised, but most of the Py_Unicode C-API expects those fields to be initialised and valid (i.e., they expect the object to be "canonical"). By Python 3.12 the legacy APIs will go away and this issue disappears: we can assume that any `Py_Unicode` object is canonical and drop the `Py_UnicodeReady` call.

This branch is here because `Py_UnicodeReady` itself can fail, as a result of memory allocation errors. It would be very hard to ever exercise this branch, but it should still be there.

Reply:
So to be clear, this branch isn't simply handling the case that `name` is not canonical: it's handling the case where `name` is not canonical *and* the attempt to make it canonical via `PyUnicode_Ready` failed. In practice, that should be a very rare occurrence indeed.